src/main/java/com/google/cloud/tools/project/AppYaml.java

```diff
@@ -16,54 +16,118 @@
 
 package com.google.cloud.tools.project;
 
-import java.io.IOException;
+import com.google.cloud.tools.appengine.api.AppEngineException;
 import java.io.InputStream;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.Collections;
 import java.util.Map;
 import javax.annotation.Nullable;
 import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.SafeConstructor;
+import org.yaml.snakeyaml.parser.ParserException;
+import org.yaml.snakeyaml.scanner.ScannerException;
 
-/** Tools for reading app.yaml */
+/** Tools for reading {@code app.yaml}. */
 public class AppYaml {
 
-  private final Map<String, ?> yamlMap;
-
+  private static final String ENVIRONMENT_TYPE_KEY = "env";
   private static final String RUNTIME_KEY = "runtime";
+  private static final String API_VERSION_KEY = "api_version";
+  private static final String APPLICATION_KEY = "application";
+  private static final String VERSION_KEY = "version";
+  private static final String SERVICE_KEY = "service";
+  private static final String MODULE_KEY = "module";
+  private static final String ENVIRONMENT_VARIABLES_KEY = "env_variables";
+
+  private final Map<String, ?> yamlMap;
 
   /**
    * Parse an app.yaml file to an AppYaml object.
    *
-   * @param appYaml the app.yaml file
-   * @throws IOException if reading app.yaml fails due to I/O errors
-   * @throws org.yaml.snakeyaml.scanner.ScannerException if reading app.yaml fails while scanning
-   *     due to malformed YAML (undocumented {@link RuntimeException} from {@link Yaml#load})
-   * @throws org.yaml.snakeyaml.parser.ParserException if reading app.yaml fails while parsing due
-   *     to malformed YAML (undocumented {@link RuntimeException} from {@link Yaml#load})
+   * @param input the input, typically the contents of an {@code app.yaml} file
+   * @throws AppEngineException if reading app.yaml fails while scanning such as due to malformed
+   *     YAML
    */
-  public AppYaml(Path appYaml) throws IOException {
-    try (InputStream in = Files.newInputStream(appYaml)) {
-      Object loaded = new Yaml().load(in);
-      if (loaded == null) {
-        yamlMap = Collections.emptyMap();
-      } else {
-        yamlMap = (Map<String, ?>) loaded;
-      }
+  @SuppressWarnings("unchecked")
+  public static AppYaml parse(InputStream input) throws AppEngineException {
+    try {
+      // our needs are simple so just load using primitive objects
+      Yaml yaml = new Yaml(new SafeConstructor());
+      Map<String, ?> contents = (Map<String, ?>) yaml.load(input);
+      return new AppYaml(contents);
+    } catch (ScannerException | ParserException ex) {
+      throw new AppEngineException("Malformed 'app.yaml'.", ex);
```

Comment:
Might be worth doing

"Malformed app.yaml: " + ex.getMessage()

depending on how informative the SnakeYaml messages are.

Reply:
They're multi-line and quite verbose.  For example:
```
while parsing a block mapping
 in 'reader', line 1, column 1:
    : m a l f o r m e d !
    ^
expected <block end>, but found ':'
 in 'reader', line 1, column 1:
    : m a l f o r m e d !
    ^
```
I'd expect that these issues would normally be surfaced to the user in some other way.